various optimizations

DanShappir opened the pull request:
I've implemented various changes that improve performance, especially in scenarios involving a lot of modules. Specifically:
1. Improved name normalization - don't do work unless it's required, e.g. breaking name into parts
2. Use a map (urlMap) to avoid performing same nameToUrl conversion multiple times
3. Moved try...catch into separate functions to enable engine optimizations
4. Reduce number of dictionary look-ups

jrburke commented:
I have not looked extensively at this yet, but some quick feedback:

* This change includes some whitespace changes, mainly some indenting changes, that are not related to the code changes, it would be good to remove those to make it more clear what code has changed.
* The urlMap entry for a module ID would need to be cleared in the case of an undef() call.

It would also be good to know if these resulted in materially measurable differences in performance. I did not think these items would show up enough to matter for slow paths in application profiling.

DanShappir commented:
1. I've fixed the spacing issues
2. I've also fixed undef() to clear urlMap
3. Regarding actual impacts: you are correct that it's small, though in a large project where files are already cached, I've seen improvements of tens of ms to a few hundred ms. I'm now also looking at checkLoaded.

emilos commented:
I've checked these changes in another project and I also see improvements. In one of the cases 570ms drops to 490ms and that seems to be quite a lot considering that these differences can be multiplied on mobile devices.
